File: src/templates/java/VectraClient.java
// Vectra gRPC client — thin idiomatic wrapper over generated stubs.
//
// Usage:
//   try (VectraClient client = new VectraClient()) {
//       List<DocumentResult> results = client.queryDocuments("my-index", "search query");
//   }
//
// Generate stubs first (see README.md).

package io.github.vectra.client;

import com.google.protobuf.util.JsonFormat;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import vectra.VectraServiceGrpc;
import vectra.VectraServiceProto.*;

import java.util.*;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

/**
 * Idiomatic Java client for the Vectra gRPC server.
 */
public class VectraClient implements AutoCloseable {
    private final ManagedChannel channel;
    private final VectraServiceGrpc.VectraServiceBlockingStub stub;

    public VectraClient() {
        this("127.0.0.1", 50051);
    }

    public VectraClient(String host, int port) {
        this.channel = ManagedChannelBuilder.forAddress(host, port)
                .usePlaintext()
                .build();
        this.stub = VectraServiceGrpc.newBlockingStub(channel);
    }

    @Override
    public void close() {
        try {
            channel.shutdown().awaitTermination(5, TimeUnit.SECONDS);
        } catch (InterruptedException e) {
            channel.shutdownNow();
            Thread.currentThread().interrupt();
        }
    }

    // ── Index Management ──────────────────────────────────

    public void createIndex(String name) {
        createIndex(name, "json", false, 512, 0);
    }

    public void createIndex(String name, String format, boolean isDocumentIndex,
                            int chunkSize, int chunkOverlap) {
        CreateIndexRequest.Builder req = CreateIndexRequest.newBuilder()
                .setIndexName(name)
                .setFormat(format)
                .setIsDocumentIndex(isDocumentIndex);
        if (isDocumentIndex) {
            req.setDocumentConfig(DocumentIndexConfig.newBuilder()
                    .setVersion(1)
                    .setChunkSize(chunkSize)
                    .setChunkOverlap(chunkOverlap)
                    .build());
        }
        stub.createIndex(req.build());
    }

    public void deleteIndex(String name) {
        stub.deleteIndex(DeleteIndexRequest.newBuilder().setIndexName(name).build());
    }

    public List<IndexInfo> listIndexes() {
        ListIndexesResponse resp = stub.listIndexes(ListIndexesRequest.getDefaultInstance());
        return resp.getIndexesList();
    }

    // ── Item Operations ───────────────────────────────────

    public String insertItem(String index, String text) {
        return insertItem(index, text, null, null, "");
    }

    public String insertItem(String index, String text, List<Float> vector,
                             Map<String, MetadataValue> metadata, String id) {
        InsertItemRequest.Builder req = InsertItemRequest.newBuilder()
                .setIndexName(index)
                .setText(text)
                .setId(id != null ? id : "");
        if (vector != null) req.addAllVector(vector);
        if (metadata != null) req.putAllMetadata(metadata);
        return stub.insertItem(req.build()).getId();
    }

    public String upsertItem(String index, String id, String text, List<Float> vector,
                             Map<String, MetadataValue> metadata) {
        UpsertItemRequest.Builder req = UpsertItemRequest.newBuilder()
                .setIndexName(index)
                .setId(id)
                .setText(text);
        if (vector != null) req.addAllVector(vector);
        if (metadata != null) req.putAllMetadata(metadata);
        return stub.upsertItem(req.build()).getId();
    }

    public ItemResult getItem(String index, String id) {
        GetItemResponse resp = stub.getItem(
                GetItemRequest.newBuilder().setIndexName(index).setId(id).build());
        return resp.hasItem() ? resp.getItem() : null;
    }

    public void deleteItem(String index, String id) {
        stub.deleteItem(DeleteItemRequest.newBuilder().setIndexName(index).setId(id).build());
    }

    public List<ItemResult> listItems(String index) {
        return listItems(index, null);
    }

    public List<ItemResult> listItems(String index, String filterJson) {
        ListItemsRequest.Builder req = ListItemsRequest.newBuilder().setIndexName(index);
        if (filterJson != null && !filterJson.isEmpty()) {
            req.setFilter(MetadataFilter.newBuilder().setFilterJson(filterJson).build());
        }
        return stub.listItems(req.build()).getItemsList();
    }

    // ── Query ─────────────────────────────────────────────

    public List<ItemResult> queryItems(String index, String text, int topK) {
        return queryItems(index, text, topK, null);
    }

    public List<ItemResult> queryItems(String index, String text, int topK, String filterJson) {
        QueryItemsRequest.Builder req = QueryItemsRequest.newBuilder()
                .setIndexName(index)
                .setText(text)
                .setTopK(topK);
        if (filterJson != null && !filterJson.isEmpty()) {
            req.setFilter(MetadataFilter.newBuilder().setFilterJson(filterJson).build());
        }
        return stub.queryItems(req.build()).getResultsList();
    }

    public List<DocumentResult> queryDocuments(String index, String query) {
        return queryDocuments(index, query, 10, 50, null, false);
    }

    public List<DocumentResult> queryDocuments(String index, String query,
                                                int maxDocuments, int maxChunks,
                                                String filterJson, boolean useBm25) {
        QueryDocumentsRequest.Builder req = QueryDocumentsRequest.newBuilder()
                .setIndexName(index)
                .setQuery(query)
                .setMaxDocuments(maxDocuments)
                .setMaxChunks(maxChunks)
                .setUseBm25(useBm25);
        if (filterJson != null && !filterJson.isEmpty()) {
            req.setFilter(MetadataFilter.newBuilder().setFilterJson(filterJson).build());
        }
        return stub.queryDocuments(req.build()).getResultsList();
    }

    // ── Document Operations ───────────────────────────────

    public String upsertDocument(String index, String uri, String text) {
        return upsertDocument(index, uri, text, "", null);
    }

    public String upsertDocument(String index, String uri, String text,
                                  String docType, Map<String, MetadataValue> metadata) {
        UpsertDocumentRequest.Builder req = UpsertDocumentRequest.newBuilder()
                .setIndexName(index)
                .setUri(uri)
                .setText(text)
                .setDocType(docType != null ? docType : "");
        if (metadata != null) req.putAllMetadata(metadata);
        return stub.upsertDocument(req.build()).getDocumentId();
    }

    public void deleteDocument(String index, String uri) {
        stub.deleteDocument(DeleteDocumentRequest.newBuilder()
                .setIndexName(index).setUri(uri).build());
    }

    public List<DocumentInfo> listDocuments(String index) {
        return stub.listDocuments(ListDocumentsRequest.newBuilder()
                .setIndexName(index).build()).getDocumentsList();
    }

    // ── Stats ─────────────────────────────────────────────

    public GetIndexStatsResponse getIndexStats(String index) {
        return stub.getIndexStats(
                GetIndexStatsRequest.newBuilder().setIndexName(index).build());
    }

    public GetCatalogStatsResponse getCatalogStats(String index) {
        return stub.getCatalogStats(
                GetCatalogStatsRequest.newBuilder().setIndexName(index).build());
    }

    // ── Lifecycle ─────────────────────────────────────────

    public HealthcheckResponse healthcheck() {
        return stub.healthcheck(HealthcheckRequest.getDefaultInstance());
    }

    public void shutdown() {
        stub.shutdown(ShutdownRequest.getDefaultInstance());
    }

    // ── Helpers ───────────────────────────────────────────

    /** Create a string metadata value. */
    public static MetadataValue metaString(String s) {
        return MetadataValue.newBuilder().setStringValue(s).build();
    }

    /** Create a numeric metadata value. */
    public static MetadataValue metaNumber(double n) {
        return MetadataValue.newBuilder().setNumberValue(n).build();
    }

    /** Create a boolean metadata value. */
    public static MetadataValue metaBool(boolean b) {
        return MetadataValue.newBuilder().setBoolValue(b).build();
    }
}

File: src/templates/csharp/VectraClient.cs
// Vectra gRPC client — thin idiomatic wrapper over generated stubs.
//
// Usage:
//   using var client = new VectraClient();
//   var results = await client.QueryDocumentsAsync("my-index", "search query");
//
// Generate stubs first (see README.md).

using System.Text.Json;
using Grpc.Net.Client;
using Vectra;

namespace Vectra.Client;

/// <summary>Idiomatic C# client for the Vectra gRPC server.</summary>
public sealed class VectraClient : IDisposable
{
    private readonly GrpcChannel _channel;
    private readonly VectraService.VectraServiceClient _stub;

    public VectraClient(string host = "127.0.0.1", int port = 50051)
    {
        _channel = GrpcChannel.ForAddress($"http://{host}:{port}");
        _stub = new VectraService.VectraServiceClient(_channel);
    }

    public void Dispose() => _channel.Dispose();

    // ── Index Management ──────────────────────────────────

    public async Task CreateIndexAsync(
        string name,
        string format = "json",
        bool isDocumentIndex = false,
        uint chunkSize = 512,
        uint chunkOverlap = 0)
    {
        var req = new CreateIndexRequest
        {
            IndexName = name,
            Format = format,
            IsDocumentIndex = isDocumentIndex,
        };
        if (isDocumentIndex)
        {
            req.DocumentConfig = new DocumentIndexConfig
            {
                Version = 1,
                ChunkSize = chunkSize,
                ChunkOverlap = chunkOverlap,
            };
        }
        await _stub.CreateIndexAsync(req);
    }

    public async Task DeleteIndexAsync(string name)
        => await _stub.DeleteIndexAsync(new DeleteIndexRequest { IndexName = name });

    public async Task<List<IndexInfo>> ListIndexesAsync()
    {
        var resp = await _stub.ListIndexesAsync(new ListIndexesRequest());
        return resp.Indexes.ToList();
    }

    // ── Item Operations ───────────────────────────────────

    public async Task<string> InsertItemAsync(
        string index,
        string text = "",
        IEnumerable<float>? vector = null,
        Dictionary<string, object>? metadata = null,
        string id = "")
    {
        var req = new InsertItemRequest
        {
            IndexName = index,
            Text = text,
            Id = id,
        };
        if (vector != null) req.Vector.AddRange(vector);
        if (metadata != null) SetMetadata(req.Metadata, metadata);
        var resp = await _stub.InsertItemAsync(req);
        return resp.Id;
    }

    public async Task<string> UpsertItemAsync(
        string index,
        string id,
        string text = "",
        IEnumerable<float>? vector = null,
        Dictionary<string, object>? metadata = null)
    {
        var req = new UpsertItemRequest
        {
            IndexName = index,
            Id = id,
            Text = text,
        };
        if (vector != null) req.Vector.AddRange(vector);
        if (metadata != null) SetMetadata(req.Metadata, metadata);
        var resp = await _stub.UpsertItemAsync(req);
        return resp.Id;
    }

    public async Task<ItemResult?> GetItemAsync(string index, string id)
    {
        var resp = await _stub.GetItemAsync(new GetItemRequest { IndexName = index, Id = id });
        return resp.Item;
    }

    public async Task DeleteItemAsync(string index, string id)
        => await _stub.DeleteItemAsync(new DeleteItemRequest { IndexName = index, Id = id });

    public async Task<List<ItemResult>> ListItemsAsync(
        string index,
        Dictionary<string, object>? filter = null)
    {
        var req = new ListItemsRequest { IndexName = index };
        if (filter != null)
            req.Filter = new MetadataFilter { FilterJson = JsonSerializer.Serialize(filter) };
        var resp = await _stub.ListItemsAsync(req);
        return resp.Items.ToList();
    }

    // ── Query ─────────────────────────────────────────────

    public async Task<List<ItemResult>> QueryItemsAsync(
        string index,
        string text = "",
        IEnumerable<float>? vector = null,
        uint topK = 10,
        Dictionary<string, object>? filter = null)
    {
        var req = new QueryItemsRequest
        {
            IndexName = index,
            Text = text,
            TopK = topK,
        };
        if (vector != null) req.Vector.AddRange(vector);
        if (filter != null)
            req.Filter = new MetadataFilter { FilterJson = JsonSerializer.Serialize(filter) };
        var resp = await _stub.QueryItemsAsync(req);
        return resp.Results.ToList();
    }

    public async Task<List<DocumentResult>> QueryDocumentsAsync(
        string index,
        string query,
        uint maxDocuments = 10,
        uint maxChunks = 50,
        Dictionary<string, object>? filter = null,
        bool useBm25 = false)
    {
        var req = new QueryDocumentsRequest
        {
            IndexName = index,
            Query = query,
            MaxDocuments = maxDocuments,
            MaxChunks = maxChunks,
            UseBm25 = useBm25,
        };
        if (filter != null)
            req.Filter = new MetadataFilter { FilterJson = JsonSerializer.Serialize(filter) };
        var resp = await _stub.QueryDocumentsAsync(req);
        return resp.Results.ToList();
    }

    // ── Document Operations ───────────────────────────────

    public async Task<string> UpsertDocumentAsync(
        string index,
        string uri,
        string text,
        string docType = "",
        Dictionary<string, object>? metadata = null)
    {
        var req = new UpsertDocumentRequest
        {
            IndexName = index,
            Uri = uri,
            Text = text,
            DocType = docType,
        };
        if (metadata != null) SetMetadata(req.Metadata, metadata);
        var resp = await _stub.UpsertDocumentAsync(req);
        return resp.DocumentId;
    }

    public async Task DeleteDocumentAsync(string index, string uri)
        => await _stub.DeleteDocumentAsync(new DeleteDocumentRequest { IndexName = index, Uri = uri });

    public async Task<List<DocumentInfo>> ListDocumentsAsync(string index)
    {
        var resp = await _stub.ListDocumentsAsync(new ListDocumentsRequest { IndexName = index });
        return resp.Documents.ToList();
    }

    // ── Stats ─────────────────────────────────────────────

    public async Task<GetIndexStatsResponse> GetIndexStatsAsync(string index)
        => await _stub.GetIndexStatsAsync(new GetIndexStatsRequest { IndexName = index });

    public async Task<GetCatalogStatsResponse> GetCatalogStatsAsync(string index)
        => await _stub.GetCatalogStatsAsync(new GetCatalogStatsRequest { IndexName = index });

    // ── Lifecycle ─────────────────────────────────────────

    public async Task<HealthcheckResponse> HealthcheckAsync()
        => await _stub.HealthcheckAsync(new HealthcheckRequest());

    public async Task ShutdownAsync()
        => await _stub.ShutdownAsync(new ShutdownRequest());

    // ── Helpers ───────────────────────────────────────────

    private static void SetMetadata(
        Google.Protobuf.Collections.MapField<string, MetadataValue> protoMap,
        Dictionary<string, object> metadata)
    {
        foreach (var (key, value) in metadata)
        {
            protoMap[key] = value switch
            {
                bool b => new MetadataValue { BoolValue = b },
                int i => new MetadataValue { NumberValue = i },
                long l => new MetadataValue { NumberValue = l },
                float f => new MetadataValue { NumberValue = f },
                double d => new MetadataValue { NumberValue = d },
                _ => new MetadataValue { StringValue = value?.ToString() ?? "" },
            };
        }
    }
}
